File: Trainings/src/lesson200407/Stack.py

class StackIterator:
    def __init__(self, head):
        self.current = head

    def __iter__(self):
        return self

    def __next__(self):
        if not self.current:
            raise StopIteration
        else:
            item = self.current.data
            self.current = self.current.next
            return item

class Node:
    def __init__(self,data=None):
        self.data = data
        self.next = None

class Stack:
    def __init__(self):
        self.head = Node()
    
    
    # def __iter__(self):
    #     return StackIterator(self.head)


    def push(self,item):
        oldfirst = self.head
        first = Node(item)
        self.head = first
        first.next = oldfirst
    
    def pop(self):
        if self.head.data == None:
            return "null"
        item = self.head.data
        self.head = self.head.next
        return item
    
    def top(self):
        if self.head.data == None:
            return "null"
        return self.head.data

    def traverse(self):
        current = self.head
        while current:
            yield current.data
            current = current.next

File: Trainings/src/lesson200407/Stack.java
package lesson200407;

import java.util.Iterator;

public class Stack  implements Iterable<String>{
	
	private static class Node {
		Node next;
		String value;
		public Node(Node next, String value) {
			this.next = next;
			this.value = value;
		}
	}

	private Node head = null;
	
	public void push(String item) {
		head = new Node(head, item);
	}
	
	public String pop() {
		if(head == null) return "null";
		String item = head.value;
		head = head.next;
		return item;
	}
	
	public String top() {
		if(head == null) return "null";
		return head.value;
	}
	
	@Override
	public String toString() {
//		StringBuilder sb = new StringBuilder();
//		Node current = head;
//		
//		while (current != null) {
//			sb.append('[').append(current.value).append("] ");
//			current = current.next;
//		}
//		
//		return sb.toString();

		StringBuilder sb = new StringBuilder();
		Iterator<String> it = iterator();
		while (it.hasNext()) {
			sb.append('[').append(it.next()).append("] ");
		}
		return sb.toString();
		
	}

	@Override
	public Iterator<String> iterator() {
		return new Iterator<String>() {
			
			Node current = head;

			@Override
			public boolean hasNext() {
				return current != null;
			}

			@Override
			public String next() {
				if (!hasNext()) return null;
				String value = current.value;
				current = current.next;
				return value;
			}
		};
	}
}
















